### PR TITLE
Apply user-defined styles via custom CSS file

### PR DESCRIPTION
**Pull request by sirlindqvist**
## Description
Load a custom CSS file in map.html to be able to change visual styles. Options reset settings button added.

## Motivation and Context
An easy way to change font/colors etc. to give the map a little more personal touch. With Google Chrome developer tools it's easy to find which elements to add to the custom.css found in static/css.

## How Has This Been Tested?
On my own map. **Thanks to all of #pr awesome Team! <3**

## Screenshots (if appropriate):
Menu with custom CSS styles:

menu

Button for reset settings back to default values:

reset

**CSS code example for above changes is included in custom.css.example.** For use, copy example file to custom.css.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by SiteEffect** [comment]
Uh, that's interesting since your preview image shows the Worldopole colors :-P @Obihoernchen

File: static/css/custom.css.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/* onoff switch layout example
+.onoffswitch-label, .onoffswitch-label:before {border-color:#256db6 !important;}
+#nav .onoffswitch-checkbox:checked+.onoffswitch-label {border-color:#256db6; background-color:#ffde4e;}
+#nav h3 {font-size:12px;}
+*/
```
Comment (Arengo) [inline static/css/custom.css.example]: Please fix to proper file ending.

Comment (sirlindqvist) [inline static/css/custom.css.example]: Done, also didn't have all values in css as the screenshot shows, now added.

File: templates/map.html [hunk 2]
```diff
@@ -360,6 +361,8 @@
             </div>
           </div>
         </div>
+          
+            <center><button class="settings" onclick="confirm('Are you sure you want to reset settings to default values?') ? (localStorage.clear(), window.location.reload()) : false">Reset</button></center>
```
Comment (sebastienvercammen) [inline templates/map.html]: Indentation is off.

File: templates/map.html [hunk 3]
```diff
@@ -34,13 +34,15 @@
...
+    <!-- custom css code-->
+    <link rel="stylesheet" type="text/css" href="static/css/custom.css">
   </head>
   <body id="top">
     <div class="wrapper">
       <!-- Header -->
       <header id="header">
         <a href="#nav"><span class="label">Options</span></a>
-        <h1><a href="#">Rocket Map</a></h1>
+        <a href="#"><img src="static/images/top-logo.png"></a>
```
Comment (sebastienvercammen) [inline templates/map.html]: We can't allow a custom logo. Stick to the "RocketMap" text (although it should actually be without space in the middle).

File: templates/map.html [hunk 4]
```diff
@@ -34,13 +34,15 @@
     <script src="{{ url_for('static', filename='js/vendor/modernizr.custom.js').lstrip('/') }}"></script>
     <!-- Toastr -->
     <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/latest/toastr.min.css">
+    <!-- custom css code-->
+    <link rel="stylesheet" type="text/css" href="static/css/custom.css">
```
Comment (sebastienvercammen) [inline templates/map.html]: Wrap some Flask/python code around it to only add this when the file exists, otherwise it will 404. I don't like errors in my console.

File: static/css/custom.css.example [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+/* menu width,padding example */
+#nav.visible {padding:10px;}
+#nav {width:250px;}
+
+/* onoff switch layout example */
+.onoffswitch-label, .onoffswitch-label:before {border-color:#256db6 !important;}
+#nav .onoffswitch-checkbox:checked+.onoffswitch-label {border-color:#256db6; background-color:#ffde4e;}
+.select2-container {font-size: 12px;}
+
+/* reset button */
```
Comment (sebastienvercammen) [inline static/css/custom.css.example]: Improve your CSS comments: capitalization, punctuation, and proper/clearer wording.

**Comment by sebastienvercammen** [comment]
Overall a good idea, but it will need to be without the logo, and you'll have to address the changes I've requested above.

And don't forget to add `custom.css` to `.gitignore`.

File: runserver.py [hunk 6]
```diff
@@ -155,6 +155,15 @@ def main():
         logging.getLogger('rpc_api').setLevel(logging.DEBUG)
         logging.getLogger('werkzeug').setLevel(logging.DEBUG)
 
+    # Check if custom.css is used otherwise fall back to default.
+    if os.path.exists(
+        os.path.join(os.path.dirname(__file__),
```
Comment (sebastienvercammen) [inline runserver.py]: #2041 introduced `root_path`, [use that instead](https://github.com/RocketMap/RocketMap/blob/develop/runserver.py#L121).

You're also not doing anything with the conditional: your HTML still contains the CSS file, and it will still error 404 if it doesn't exist. You should only add the CSS if it exists, via the templating in the HTML.

File: runserver.py [hunk 7]
```diff
@@ -155,6 +155,15 @@ def main():
...
+                     'static/css/custom.css')):
+        log.info('custom.css found, applying user-defined settings.')
+    else:
+        log.info('No custom.css found, using default settings.')
+        pass
```
Comment (sebastienvercammen) [inline runserver.py]: `pass` doesn't do anything: https://docs.python.org/2/tutorial/controlflow.html#pass-statements

File: templates/map.html [hunk 8]
```diff
@@ -360,6 +360,9 @@
             </div>
           </div>
         </div>
+      <div style="margin-top:10px;">
+        <center><button class="settings" onclick="confirm('Are You sure You want to reset settings to default values?') ? (localStorage.clear(), window.location.reload()) : false" style="padding:0px 10px 0px 10px;"><i class="fa fa-refresh" aria-hidden="true"></i>&nbsp; Reset</button></center>
```
Comment (sebastienvercammen) [inline templates/map.html]: Fix capitalization:

> Are You sure You

and clarify the button's text: "Reset" is not as clear as "Reset Settings".

File: templates/status.html [hunk 9]
```diff
@@ -73,7 +73,7 @@
   </div>
 </nav>
 <div class="spacing"></div>
-<div class="status_container" id="status_container">
+<div class="status_container" id="status_container" style="margin-top:10px;">
```
Comment (sebastienvercammen) [inline templates/status.html]: Why add inline styling if we have both a class and an ID to use in our CSS files?

Comment (sirlindqvist) [inline templates/status.html]: This is because there is no padding between top and the hash key table if viewing on a desktop. So if users don't wanna use a custom css file there will be no space between so i think this would be default.

File: templates/map.html (hunk 8 above)
Comment (sirlindqvist) [inline templates/map.html]: Thought 'You' should be used but there's no problems changing to 'you' :) Will also fix the button text.

File: runserver.py (hunk 6 above)
Comment (sirlindqvist) [inline runserver.py]: I'll check this.

File: runserver.py (hunk 7 above)
Comment (sirlindqvist) [inline runserver.py]: Looking at this too.

**Comment by SiteEffect** [comment]
Could you please add your sample .css here in the description or at least PM me with it?

File: templates/status.html (hunk 9 above)
Comment (sebastienvercammen) [inline templates/status.html]: @sirlindqvist Everything is 👍 for merge, but the inline CSS needs to be moved to the separate CSS files. Use the class names and IDs to add the margin, without using inline CSS.

**Comment by Alderon86** [comment]
@elincognito solved with last commit :)

**Comment by elincognito** [comment]
@Alderon86 no it isn't :D

**Comment by Alderon86** [comment]
@elincognito your right commit was messed up, wait a min or 2 :D

**Comment by Alderon86** [comment]
there we go :)

**Comment by sirlindqvist** [comment]
@elincognito now it is :)

**Comment by elincognito** [comment]
@sirlindqvist :+1:

**Comment by friscoMad** [comment]
Please add it also for all other pages: stats, status and mobile.
Also adding a custom class in each page body (for example: map class in body tag) could help to manage all styles in one file.

File: templates/map.html [hunk 10]
```diff
@@ -34,19 +34,24 @@
...
+      {% endif %}
   </head>
-  <body id="top">
+  <body id="top" class="mapPage">
     <div class="wrapper">
       <!-- Header -->
       <header id="header">
         <a href="#nav"><span class="label">Options</span></a>
-        <h1><a href="#">Rocket Map</a></h1>
+        <h1><a href="./">RocketMap &nbsp;<i class="fa fa-rocket" aria-hidden="true"></i></a></h1>
```
Comment (sebastienvercammen) [inline templates/map.html]: Don't use `&nbsp;` for spacing, unless you specifically need a non-breaking space.

File: templates/map.html [hunk 11]
```diff
@@ -34,19 +34,24 @@
...
         <a href="#stats" id="statsToggle" class="statsNav" style="float: right;"><span class="label">Stats</span></a>
+        <a href="status" target="_blank" style="float: right;"><span class="label">Status</span></a>
       </header>
       <!-- NAV -->
       <nav id="nav">
         <div id="nav-accordion">
-          <h3>Marker Settings</h3>
+          <h3>Marker Settings &emsp;<i class="fa fa-map-marker" aria-hidden="true"></i></h3>
```
Comment (sebastienvercammen) [inline templates/map.html]: Same as above, why was `&emsp;` added for spacing?

Comment (sirlindqvist) [inline templates/map.html]: How do you want me to do? To control this with CSS is not an option here since everyone won't use it. I wanted to add some font-awesome icons to the menu.

Comment (friscoMad) [inline templates/map.html]: You can add FA icons without &emsp; or &nbsp; just leave <i...</i>and delete those spaces

File: static/sass/layout/_header.scss [hunk 12]
```diff
@@ -72,6 +72,20 @@
         margin-left: 0.5em;
       }
 
+      &[href="status"] {
```
Comment (friscoMad) [inline static/sass/layout/_header.scss]: Why not just adding the FA icon to the links?

Comment (sirlindqvist) [inline static/sass/layout/_header.scss]: i can but i dont know howto make spaces since &nbsp; is not allowed.

Comment (Alderon86) [inline static/sass/layout/_header.scss]: @friscoMad why should we move it over to the links directly? this jzst follows the codestyle, a rewrite of that would be a different pr in my eyes

Comment (friscoMad) [inline static/sass/layout/_header.scss]: I hadn't seen that all other icons are setted in css, it is not the standard and while I admit that icons are totally visual issue and that should be handled in css over html, FA is usually added in the html as content, thats how the library tells you to use it and it is where you expect it to be.

The to be consistent with the codebase we would need to move all rocket icons to css too.

File: templates/map.html (hunk 11 above)
Comment (Alderon86) [inline templates/map.html]: in this case without a pre-defined space its looking quite ugly, but pleaser correct me if theres a better / simpler soloution

File: templates/status.html [hunk 13]
```diff
@@ -39,11 +39,10 @@
 <!-- Header -->
 <header id="header">
     <a href="#nav"><span class="label">Options</span></a>
-    <h1><a href="./">RocketMap &nbsp;<i class="fa fa-rocket" aria-hidden="true"></i></a></h1>
+    <h1><a href="#">RocketMap<i class="fa fa-rocket fa-fw"></i></a></h1>
```
Comment (sebastienvercammen) [inline templates/status.html]: Any reason you reverted from `./` back to `#`?

Comment (sirlindqvist) [inline templates/status.html]: Yepp :) I wanted the status page to be a separate page. When icon is clicked you get it in a new tab so linking back to the map is not what i want here. Stats page on the other side is "more a part of the map" so its not opened in a new tab and clicking RocketMap banner brings you back to main map. Thats how i see "the flow" :)
